zkvm/src/predicate.rs

```diff
@@ -181,75 +151,195 @@ impl Into<CompressedRistretto> for Predicate {
 
-        // dummy predicates
-        let preds = vec![
-            Predicate::Opaque(gens.B.compress()),
-            Predicate::Opaque(gens.B_blinding.compress()),
-        ];
+impl CallProof {
+    pub fn serialized_length(&self) -> usize {
+        // VerificationKey is a 32-byte array
+        // MerkleNeighbor is a 32-byte array
+        32 + 4 + self.neighbors.len() * 32
+    }
 
-        let pred = Predicate::disjunction(preds.clone()).unwrap();
-        let op = pred.prove_disjunction(&preds);
-        assert!(op.verify().is_ok());
+    pub fn parse(data: &[u8]) -> Result<Self, VMError> {
+        SliceReader::parse(data, |r| {
+            let mut call_proof = CallProof::default();
+            call_proof.verification_key = VerificationKey(r.read_point()?);
+            let positions = r.read_u32()?;
+            let mut neighbors = vec![];
+            if r.len() % 32 != 0 {
+                return Err(VMError::FormatError);
+            }
+            let num_neighbors = r.len() / 32;
+            for i in 0..num_neighbors {
+                let neighbor_bytes = r.read_u8x32()?;
+                if positions & (1 << i) == 0 {
+                    neighbors.push(MerkleNeighbor::Left(neighbor_bytes));
+                } else {
+                    neighbors.push(MerkleNeighbor::Right(neighbor_bytes));
+                }
+            }
+            call_proof.neighbors = neighbors;
+            Ok(call_proof)
+        })
     }
 
-    #[test]
-    fn valid_disjunction2() {
-        let gens = PedersenGens::default();
+    /// Serializes the call proof to a byte array
+    pub fn encode(&self, buf: &mut Vec<u8>) {
+        encoding::write_point(&self.verification_key.0, buf);
+
+        let mut positions: u32 = 0;
+        for (i, n) in self.neighbors.iter().enumerate() {
+            match n {
+                MerkleNeighbor::Right(_) => {
+                    positions = positions | (1 << i);
+                }
+                _ => {}
+            }
+        }
+        encoding::write_u32(positions, buf);
+        for n in &self.neighbors {
+            match n {
+                MerkleNeighbor::Left(l) => encoding::write_bytes(l, buf),
+                MerkleNeighbor::Right(r) => encoding::write_bytes(r, buf),
+            }
+        }
+    }
+}
+
+impl PredicateLeaf {
+    /// Downcasts the predicate leaf to a program.
+    pub fn to_program(self) -> Result<Program, VMError> {
+        match self {
+            PredicateLeaf::Program(p) => Ok(p),
+            _ => Err(VMError::TypeNotProgram),
+        }
+    }
+}
+
+impl MerkleItem for PredicateLeaf {
+    fn commit(&self, t: &mut Transcript) {
+        match self {
+            PredicateLeaf::Program(prog) => {
+                let mut buf = Vec::new();
+                prog.encode(&mut buf);
+                t.commit_bytes(b"predicate_leaf", &buf);
+            }
+            PredicateLeaf::Blinding(bytes) => {
+                t.commit_bytes(b"predicate_leaf", &bytes.clone());
```

Comment:
see the comment above
```suggestion
                t.commit_bytes(b"blinding", &bytes.clone());
```